### DISCUSSION TITLE
Result processor for results from raw SQL statement

### DISCUSSION THREAD

**Discussion by jhslashdb** [sqlalchemy/sqlalchemy#11585] [Usage Questions]
Have a question regarding result processors and raw SQL statements.  When using ORM with mapped classes, I extend SQL types with custom result processors etc.  However, if I execute a raw SQL statement, it appears that SQLAlchemy identified all the columns as NullType, and AFAIK you cannot extend that type (and even if you can, not sure if it would be a good idea).  Is there a recommended way to create a result processor for such cases?

**Comment by zzzeek** [comment]
yes use the text().columns() method

https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#getting-orm-results-from-textual-statements

https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.TextClause.columns

**Reply by jhslashdb** [reply]
OK, that makes sense.  What about a scenario where the column names weren't known in advance.  Is there anything that can be done in that case?

**Reply by zzzeek** [reply]
the DBAPI will correctly discern between ints, strings, and usually dates (not on sqlite) but if you need more information than that you'd need to know the types in advance.

**Reply by jhslashdb** [reply]
One last question  - in SQLAlchemy 1.x, I was able to execute a query, take the result object, and point to a custom result processor function for each row using the `result._metadata._processors` list.  This doesn't work in SQLAlchemy 2.x;  it seems like the result processors are cached early on.  Is there another object in the result that I could use for pointing a custom result processor?

**Reply by zzzeek** [reply]
well when you were patching your own `_metadata._processors` you were hacking through the internals.   Sure, you can do something like that here as well, `_metadata._processors` should be mostly the same compared to 1.4, but overall we dont have a public API to replace the processors for a statement after it was executed.

it would be a lot more straightforward if you took the result that you got, then ran simple Python functions on the row received to do additional translations you're looking for after the fact.
